### ISSUE TITLE
Bundle version deadlock installing 6.3.1

### ISSUE THREAD

**Issue by fjorba** [rism-ch/muscat#967]
I've tried to install 6.3.1 on the same Debian Buster (10.5) that I've successfully installed all my previous Muscat versions before.  However, I cannot find a solution to go ahead with the installation due to Bundler conflicting versions.

Current https://github.com/rism-ch/muscat/blob/develop/2-%20INSTALL.rdoc instructions require a bundler less than 2.0.

<pre>
sudo gem install bundler -v '< 2.0' ## force 1.7 until we get compatibile with 2.0
</pre>

Debian Buster provides the 1.17 version by default, and a backport for 2.1.4.  I've tried both, and installation break at different points.

With 1.17, this is what I found:

<pre>
~/muscat/muscat-6.3.1$ dpkg -l | grep bundler
ii  ruby-bundler        1.17.3-3+deb10u1      all    Manage Ruby application dependencies (runtime)

~/muscat/muscat-6.3.1$ bundle install --path vendor/bundle
Traceback (most recent call last):
        2: from /usr/bin/bundle:23:in `<main>'
        1: from /usr/lib/ruby/2.5.0/rubygems.rb:303:in `activate_bin_path'
/usr/lib/ruby/2.5.0/rubygems.rb:284:in `find_spec_for_exe': Could not find 'bundler' (2.1.4) required by your ~/muscat-6.3.1/Gemfile.lock. (Gem::GemNotFoundException)
To update to the latest version installed on your system, run `bundle update --bundler`.
To install the missing version, run `gem install bundler:2.1.4`

~/muscat/muscat-6.3.1$ grep --color bundler Gemfile*
Gemfile.lock:      bundler (>= 1.3.0)
</pre>

I must confess that I cannot understand the message; with my basic Ruby ecosystem knowledge, I don't see the 2.1.4 requirement in the Gemfile.

However, I installed the 2.1.4 backport:

<pre>
# apt install -t buster-backports ruby-bundler

~/muscat/muscat-6.3.1$ dpkg -l | grep bundler
ii  ruby-bundler        2.1.4-2~bpo10+1       all    Manage Ruby application dependencies (runtime)
~/muscat/muscat-6.3.1$ bundle install --path vendor/bundle
[DEPRECATED] The `--path` flag is deprecated because it relies on being remembered across bundler invocations, which bundler will no longer do in future versions. Instead please use `bundle config set path 'vendor/bundle'`, and stop using this flag
Fetching gem metadata from https://rubygems.org/............
Fetching https://github.com/rism-ch/blacklight
[...]
Bundle complete! 60 Gemfile dependencies, 160 gems now installed.
Bundled gems are installed into `./vendor/bundle`
Post-install message from i18n:
[...]

[Configure db, etc]

~/muscat/muscat-6.3.1$ bundle exec rake db:migrate
rake aborted!
TypeError: superclass mismatch for class Command
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/thor-1.0.1/lib/thor/command.rb:2:in `<class:Thor>'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/thor-1.0.1/lib/thor/command.rb:1:in `<top (required)>'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/thor-1.0.1/lib/thor/base.rb:1:in `require_relative'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/thor-1.0.1/lib/thor/base.rb:1:in `<top (required)>'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/thor-1.0.1/lib/thor/group.rb:1:in `require_relative'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/thor-1.0.1/lib/thor/group.rb:1:in `<top (required)>'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/railties-5.2.4.3/lib/rails/generators.rb:6:in `require'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/railties-5.2.4.3/lib/rails/generators.rb:6:in `<top (required)>'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/coffee-rails-4.2.2/lib/coffee/rails/engine.rb:2:in `require'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/coffee-rails-4.2.2/lib/coffee/rails/engine.rb:2:in `<top (required)>'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/coffee-rails-4.2.2/lib/coffee-rails.rb:2:in `require'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/coffee-rails-4.2.2/lib/coffee-rails.rb:2:in `<top (required)>'
/usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:74:in `require'
/usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:74:in `block (2 levels) in require'
/usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:69:in `each'
/usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:69:in `block in require'
/usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:58:in `each'
/usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:58:in `require'
/usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler.rb:174:in `require'
~/muscat-6.3.1/config/application.rb:8:in `<top (required)>'
~/muscat-6.3.1/Rakefile:4:in `require'
~/muscat-6.3.1/Rakefile:4:in `<top (required)>'
~/muscat-6.3.1/vendor/bundle/ruby/2.5.0/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
(See full trace by running task with --trace)
</pre>

Adding --trace to the last command didn't change the output.

I haven't found any of those problems with the previous versions, up to and including 6.2.

**Comment by xhero** [comment]
oh! I think there is a bug in the docs, muscat should run with 2.0 (we use 2.1.4), can you try to run just

 gem install bundler

this should get 2.1.4 directly without passing from debian. What version of ruby are you using? I recommend a recent one.

**Comment by fjorba** [comment]
I'm using current Debian defaults as much as I can; current Debian stable Ruby is 2.5.5.  If I upgrade the system wide bundler installing backported 2.14 it should also work , the same way that with `gem install bundler`, and I avoid to do it for my three or four Mustcat test installations.

Anyway, using 2.1.4, I get this `TypeError: superclass mismatch for class Command`.  Do you thing that it is Muscat related, or due to something in my system versions?  I don't know how to continue.

**Comment by xhero** [comment]
That error does not seem directly related to the Muscat code, did you update the gems in Muscat too? We upgraded to rails 5.2

**Comment by fjorba** [comment]
Yes, I installed anew, in a new directory, following the instructions.  I understand that `bundle install --path vendor/bundle` and reading the downloading logs proves that everything is up to date.

However, reading again my original message, I suspect that the error could be related to rake, as it is the last module referenced.  Debian has 12.3.1, but in the logs it appears that Muscat installation has taken  a local 13.0:

https://packages.debian.org/rake

Which version are you using?

**Comment by fjorba** [comment]
Google has found me this information regading the same message on a Redmine upgrade:

https://www.redmine.org/boards/2/topics/59737

**Comment by fjorba** [comment]
Playing with different versions of rake and thor, I'm getting different error messages, but I suspect that pinning the right versions of those two gems would help to bootstrap 6.3.1.

**Comment by fjorba** [comment]
After googling a while, I've found this workaround; I don't know how serious it is, but maybe it helps you to understand better the issue:

<pre>
diff --git a/Gemfile b/Gemfile
index 77c845e..8d85421 100644
--- a/Gemfile
+++ b/Gemfile
@@ -150,6 +150,9 @@ gem "recaptcha"
 # To render markdown
 gem 'redcarpet'

+# https://github.com/erikhuda/thor/issues/721
+gem 'thor', '~> 0.20.3'
+
 group :test do
   gem "rspec"
   gem 'rspec-rails', '~> 3.5'
diff --git a/Gemfile.lock b/Gemfile.lock
index 3ab97b8..1648427 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -416,7 +416,7 @@ GEM
     sqlite3 (1.4.2)
     terrapin (0.6.0)
       climate_control (>= 0.0.3, < 1.0)
-    thor (1.0.1)
+    thor (0.20.3)
     thread_safe (0.3.6)
     tilt (2.0.10)
     turbolinks (5.2.1)
@@ -500,6 +500,7 @@ DEPENDENCIES
   sqlite3
   sunspot_rails!
   sunspot_solr!
+  thor (~> 0.20.3)
   turbolinks (~> 5.2.1)
   uglifier (~> 4.2)

diff --git a/config/initializers/recaptcha.rb b/config/initializers/recaptcha.rb
index 5b33385..bc7afc6 100644
--- a/config/initializers/recaptcha.rb
+++ b/config/initializers/recaptcha.rb
@@ -1,5 +1,5 @@
 Recaptcha.configure do |config|
-    config.site_key  = Rails.application.credentials.recaptcha[:site_key]
-    config.secret_key = Rails.application.credentials.recaptcha[:secret_key]
+#    config.site_key  = Rails.application.credentials.recaptcha[:site_key]
+#    config.secret_key = Rails.application.credentials.recaptcha[:secret_key]
 #    config.proxy = 'http://myproxy.com.au:8080'
-end
\ No newline at end of file
+end
</pre>

I've commented out this recaptcha credentials, because it seems that I'm missing some RAILS_MASTER_KEY value:

```
$ rails console
Loading development environment (Rails 5.2.4.3)
irb(main):001:0> Rails.application.credentials
=> #<ActiveSupport::EncryptedConfiguration:0x000055efd8b98660 
@key_path=#<Pathname:/home/ifmuc/muscat/muscat-6.3.1/config/master.key>, 
@content_path=#<Pathname:/home/ifmuc/muscat/muscat-6.3.1/config/credentials.yml.enc>, 
@env_key="RAILS_MASTER_KEY", @raise_if_missing_key=false, @config={}, @options={}>
```

**Comment by fjorba** [comment]
The versions conflict is solved (or, at least, I found a workaround in #972). The recaptcha is another issue, maybe because I misread the documentation, or it was not clear enough. Anyway, I'm closing this issue.
